test/spec.yml

```diff
@@ -1373,3 +1374,346 @@
+          "curl -H X-Binaris-Api-Key:* https://*"
+    - in: bn trigger upsert testtriggerbadsourceuri --source-uri "redis+tls://:${REDIS_PASS}@misc-redis.binaris.io/0#numbers" --target-func testtriggerbadsourceurifunc
+      out: CREATED
+    - in: bn trigger upsert testtriggerbadsourceuri --source-uri redis://not-a-valid-redis-uri --target-func testtriggerbadsourceurifunc
+      exit: 1
+      err: |-
+        RequestId: *
+        Internal Binaris server error*
+
+- test: Test new trigger with missing target function (bad-path)
+  steps:
+    - in: bn trigger upsert notexistingfunctrigger --source-uri "redis+tls://:${REDIS_PASS}@misc-redis.binaris.io/0#numbers" --target-func nonexistingfunc
+      exit: 1
+      err: |-
+        RequestId: *
+        Error: No such function
+
+- test: Test update trigger with missing target function (bad-path)
+  cleanup:
+    - bn trigger remove notexistingfuncupdatetrigger
+    - bn remove notexistingfuncupdatetriggerfunc
+  steps:
+    - in: bn create node8 notexistingfuncupdatetriggerfunc
+      out: |-
+          Created function * in /home/dockeruser/test
+            (use "bn deploy notexistingfuncupdatetriggerfunc" to deploy the function)
+    - in: bn deploy notexistingfuncupdatetriggerfunc
+      out: |-
+        Deployed function notexistingfuncupdatetriggerfunc
+        Invoke with one of:
+          "bn invoke notexistingfuncupdatetriggerfunc"
+          "curl -H X-Binaris-Api-Key:* https://*"
+    - in: bn trigger upsert notexistingfuncupdatetrigger --source-uri "redis+tls://:${REDIS_PASS}@misc-redis.binaris.io/0#numbers" --target-func notexistingfuncupdatetriggerfunc
+      out: CREATED
+    - in: bn trigger upsert notexistingfuncupdatetrigger --source-uri "redis+tls://:${REDIS_PASS}@misc-redis.binaris.io/0#numbers" --target-func nonexistingfunc
+      exit: 1
+      err: |-
+        RequestId: *
+        Error: No such function
+
+- test: Test new trigger when it already exists with different name (bad-path)
+  cleanup:
+    - bn trigger remove testduplicatetrigger
+    - bn remove testduplicatetriggerfunc
+  steps:
+    - in: bn create node8 testduplicatetriggerfunc
+      out: |-
+          Created function * in /home/dockeruser/test
+            (use "bn deploy testduplicatetriggerfunc" to deploy the function)
+    - in: bn deploy testduplicatetriggerfunc
+      out: |-
+        Deployed function testduplicatetriggerfunc
+        Invoke with one of:
+          "bn invoke testduplicatetriggerfunc"
+          "curl -H X-Binaris-Api-Key:* https://*"
+    - in: bn trigger upsert testduplicatetrigger --source-uri "redis+tls://:${REDIS_PASS}@misc-redis.binaris.io/0#numbers" --target-func testduplicatetriggerfunc
+      out: CREATED
+    - in: bn trigger upsert testduplicatetriggerthatwillnotbecreated --source-uri "redis+tls://:${REDIS_PASS}@misc-redis.binaris.io/0#numbers" --target-func testduplicatetriggerfunc
+      exit: 1
+      err: |-
+        RequestId: *
+        Bad Request
```

Comment:
Not the best error string, we should say that a trigger already exists for source / target combination.

Reply:
yet it passes...
this needs to be updated on `spice`

Reply:
Check the error returned by spice, it should contain the reason, this is bad UX.